### PR TITLE
dev/wordpress#147 Reworking output of Wordpress Access control table

### PR DESCRIPTION
**Pull request by shaneonabike**
Overview
----------------------------------------
As described in [dev/wordpress#147](https://lab.civicrm.org/dev/wordpress/-/issues/147 ) the idea is to rework the CiviCRM Wordpress Acess control page. I wanted to make it easier for people to be able to scroll and still see the Roles that they are assigning much like Drupal does.

Before
----------------------------------------

+ Scrolling causes the roles to be lost
+ Users need to scroll back up to see which Roles are assigned to what column in the table

Selection_001

After
----------------------------------------

+ Added the roles into a ```thead``` column using the new HTML5 formatting and the rest of the data in the ```tbody```
+ Added additional CSS tags to the ```thead``` to allow for sticky to apply when scrolling

Selection_002

Technical Details
----------------------------------------

+ I added a new CSS style called ```crm-wordpress-permissions``` because I wasn't sure if we wanted this to be global. I do see that this could create glut though. I'm open to any suggestions and if people just want me to apply this to a sticky tag I can.

### CODE REVIEW

**Comment by aydun** [comment]
FYI @vingle

**Comment by vingle** [comment]
As this is specific to WordPress I wonder if the CSS change should go in the related package's CSS file rather than the global civicrm.css? https://github.com/civicrm/civicrm-wordpress/blob/master/assets/css/civicrm.admin.css

TBH - I've no idea why joomla.css ships with CiviCRM core while the WordPress css file is in the WP package. Perhaps @christianwach has an idea?

**Comment by shaneonabike** [comment]
Either or. I thought it should go into the CiviCRM b/c that's where I located the sticky/table mods, but I also agree it makes sense to put it into the Wordpress. If you want me to move it and create a new PR let me know I'm happy to do that.

I have tested this one

- [x] Edge
- [x] Firefox
- [x] Chrome

**Comment by vingle** [comment]
@shaneonabike for me it's a new thing to be putting css specific to only one CMS inside civicrm.css. One of the advantages of keeping it separate is if/when this giant css file is replaced, it won't be lost. 

On the other hand, going back to your earlier question - _why shouldn't sticky headers on scroll be global with the right class?_ - I'm a fan of making it global (am assuming that's what the goal of .sticky`/`.sticky-header` is). What say you @artfulrobot?

**Comment by shaneonabike** [comment]
@vingle If you want to make this global I'm happy to do that. I definitely agree about separating things per CMS that makes sense. Furthermore, I prefer _not_ to make CMS specific stuff if I don't have to. I saw there was other sticky classes but it seemed strange to me, because it's like it's initiated by JQuery (doesn't have the same CSS attributes I have).

How about I just rework this to be a bit more global and if we decide to roll this across all the CRM then we can slowly picky away at changing the old ```sticky``` tags. We should be using the CSS relevant stuff when we can anyway

Give me your thoughts and I'll change up the code and then we can discuss creating a new merge to migrate this across the entire CRM

cc @vingle @artfulrobot

**Comment by christianwach** [comment]
> I've no idea why joomla.css ships with CiviCRM core while the WordPress css file is in the WP package.

The WordPress `civicrm.admin.css` stylesheet is only used on CiviCRM admin subpages, i.e. those which the CiviCRM-WordPress plugin supplies. It could be loaded in CiviCRM admin more generally, I suppose, but then it would perhaps be confusing to have the WordPress Access Control table in CiviCRM-Core and its styles in CiviCRM-WordPress.

If it were my decision (it's not) I'd have *everything* WordPress-related in CiviCRM-WordPress - including the WordPress Access Control table screen and its styles.

Having got that off my chest (phew) I support the idea of having a "one true sticky header" class. Makes sense that tables are the next thing to tackle now that discovery tabs have been dealt with.

**Comment by vingle** [comment]
@christianwach - it does seem sensible to keep CMS specific items somewhere specific to that CMS. I guess the question is if it should be part of civicrm core in `/css/` like `backdrop.css`, `menubar-wordpress.css` and `joomla.css` (90% of which we [recently ditched](https://github.com/civicrm/civicrm-core/commit/73a97f842fef9ace6c8284404d0aee8994cffef1)),  or in the CMS specific extension, ie `.civicrm.admin.css` which makes some sense, but also forces this issue to be two PRs in two places (and is something Backdrop/Joomla/Drupal doesn't do, afaik). So yes - I back putting all CMS stuff together in one identifiable place (wherever that is).

Am aware we're stumbling into a bigger discussion. This is currently on my mind because of trying to compile CMS resets that are useful for any CMS recently. E.g. the default Claro theme on Drupal 9+ requires dozens of css resets because it adds opinionated styling to a bunch of base HTML and shared classnames (sets `<td>` to be 2rem tall, for e.g). But those resets are also good defense for other CMS admin themes who may one day try to do something similar. So I'm inclined to think they should be on all CMS builds (other than Standalone I guess!). TLDR ideally as much is universal as possible.

Sidenote in the context of getting things off chest, tables really are something aren't they. So far I've counted:

- table
- table.dataTable / table.dataTable.hover / table.dataTable.display / table.dataTable.order-column.stripe
- table.row-highlight 
- table.selector
- table.form-layout-compressed
- table.report
- table#map-field-table
- table.table-condensed / table.table-bordered / table.table-striped (all Bootstrap)

And then rows inside these could be any of `tr.even`/`tr.odd`, `tr.even-row`/`tr.odd-row`, `tr.crm-row-selected`, `.tr-columnheader-dark`, `tr.crm-child-row`, `tr.crm-row-child` (!).

And to top it off Greenwich and Shoreditch also style tables contextually - whether there is a border or not around it, say, can be as much about where it is found in Civi as the classes above.

**Comment by vingle** [comment]
> How about I just rework this to be a bit more global and if we decide to roll this across all the CRM then we can slowly picky away at changing the old sticky tags. We should be using the CSS relevant stuff when we can anyway

Yes to this @shaneonabike – sounds a great idea, a goal of deprecating sticky headers with JQuery in favor of `position:sticky` css.

**Comment by christianwach** [comment]
> So far I've counted:

@vingle 😱 

> How about I just rework this to be a bit more global

@shaneonabike Anything you can do to move things towards a more sane situation would be brilliant.

**Comment by christianwach** [comment]
> it does seem sensible to keep CMS specific items somewhere specific to that CMS. I guess the question is if it should be part of civicrm core in /css/ like backdrop.css, menubar-wordpress.css and joomla.css (90% of which we [recently ditched](https://github.com/civicrm/civicrm-core/commit/73a97f842fef9ace6c8284404d0aee8994cffef1)), or in the CMS specific extension, ie .civicrm.admin.css which makes some sense, but also forces this issue to be two PRs in two places

@vingle Yes, good summary; that's what I was trying to articulate as a possible source of confusion.

As I said, in an ideal world, I would move everything WordPress-related (code, templates and styles) to the CiviCRM-WordPress repo. I think CIviCRM-Core would benefit from this too as it could then concentrate on implementing standalone functionality. But, yeah, that's probably a longer discussion for somewhere other than this PR 😄

**Comment by vingle** [comment]
> in an ideal world, I would move everything WordPress-related (code, templates and styles) to the CiviCRM-WordPress repo.

It would make a lot of sense. I'm guessing there's a constraint in how each CMS loads resources (or knowledge around that) that forces the current mixed methods. Is it only CSS that isn't in the Civi-WP repo?

**Comment by colemanw** [comment]
> @vingle I'm guessing there's a constraint in how each CMS loads resources (or knowledge around that) that forces the current mixed methods.

A lot of people who contribute to CiviCRM don't have the complete picture (understandably, it's an enormous codebase), so they wind up sticking things in not-the-perfect place. That's usually the explanation for "why is this code here and not there?"

> @christianwach If it were my decision (it's not)

It is! It can be! I fully support cleanup efforts that make Civi more sane, as does the rest of the core team. Anything you want to do along those lines we'll try to help with review/merge of your PRs.

**Comment by christianwach** [comment]
> Is it only CSS that isn't in the Civi-WP repo?

@vingle No, there's tons of stuff in CiviCRM-Core that's CMS-related. It would be a huge task to tease it all out into the relevant CMS-specific repos.

Ideally wherever there's an allowance made in core for a CMS, there could be a Symfony event fired which the CMS-specific code can listen to and use to modify the standalone defaults. Templates can be overridden, classes can be added and extended, scripts and styles can be added. It's all _possible_...

> It is! It can be!

@colemanw Like I said, in an ideal world I'd have the time and resources to do that - and perhaps an extra few dozen hours in the day 😆

**Comment by vingle** [comment]
Yes to 2, sounds great.

As for 1, I think it needs a broader discussion because – for css at least – Joomla, Backdrop and Drupal are all in CiviCRM-core and that's unlikely to change soon. Am aware I'm contradicting what I said above, but there's arguments on both sides. Unless/until all CMS specific resources are moved to their respective packages, then it doesn't necessarily make things tidier to do that just for WP, it just deepens the different patterns (and the main benefit for the effort of doing this is to 'make things tider'/more maintainable). In fact the shortest path to make things more consistent/tidy would be to move civicrm.admin.css from wp-civicrm to civicrm-core as `/css/wordpress.css`.

**Comment by shaneonabike** [comment]
@christianwach @vingle Do you think it would be realistic to accept these changes, and then we can work on the other fixes after?

File: css/civicrm.css [hunk 1]
```diff
@@ -19,6 +19,14 @@
...
 }
 
+/* CMS specific padding for sticky header padding */
+.wp-toolbar {
+  --crm-sticky-top: 40px;
+}
+.wp-toolbar .crm-menubar-below-cms-menu {
+  --crm-sticky-top: 72px;
+}
+
```
Comment (vingle) [inline css/civicrm.css:29]: If these are the only classes that define this variable, this will mean `.crm-sticky thead` below won't have a value for other CMSs than WordPress? ie it will only work on WordPress? (which would obvs be fine if this was all in the WP package)

Comment (christianwach) [reply css/civicrm.css:29]: @shaneonabike Also don't forget to accommodate responsiveness, i.e. the WordPress admin menu becomes 48px tall below a certain breakpoint.

Comment (shaneonabike) [reply css/civicrm.css:29]: @christianwach right good point. I'll test that and add the proper responsive sizes.

@vingle that's correct. At present this new tag is only used in the Wordpress page. My idea was to roll this out, and then in the other ticket I created add the corresponding sizes for the other systems. I can't really handle Joomla but I can do Drupal (since we have that build here). But I wanted to slowly roll it out to tackle this one problem, and then grow it accordingly... does that sound reasonable?

Short version: Right now this new tag I created isn't being used anywhere except the Wordpress Access table. Once we create a merge request for the rest it could include the new sizes for Drupal/Joomla/you name it

Comment (vingle) [reply css/civicrm.css:29]: @shaneonabike - then I'm afraid I think it's better this goes in the WordPress package's civicrm.admin.css file. 

At the moment there aren't any CMS-specific overrides in civicrm.css (or maybe one - that shouldn't be there) and I'd rather not plan to add a series of them, given there's a plan to to replace civicrm.css.

Do you know what other tables use `.crm-sticky` and `sticky-header`? Is it possible it's just the WordPress Permissions table? I can't picture other instances.

Comment (christianwach) [reply css/civicrm.css:29]: @vingle The `civicrm.admin.css` file doesn't load on "standard" CiviCRM admin screens, so it can't go in there.

Comment (christianwach) [reply css/civicrm.css:29]: @vingle No, there's no equivalent as things stand. The CiviCRM-WordPress package [adds some admin pages](https://github.com/civicrm/civicrm-wordpress/pull/231) that are specific to WordPress. That CSS file is loaded there and for the Dashlet.

Comment (shaneonabike) [reply css/civicrm.css]: Hi folks,

I think that I am going to make everyone happy! I just noticed that there is a JS function that basically determines whether to position the actual CiviCRM menu. In doing so, it drops a ```--crm-menubar-bottom``` inline to basically indicate the bottom of the menu bar. I changed the code to use that since it's 100% responsive as I scale the window. Also it works for Drupal (probably Joomla but I don't have an instance to test).

Then at ```782px``` the menu gets fixed to the top (same in Drupal). So I added a media catch for that and fixed the heights. The Drupal class isn't perfect but should work since there is no class specifically saying there's an admin so I went for whether the menubar is there or not.

@vingle Sticky is being used (the old way) in several areas if you grep that. Mostly, legacy search but also ```Grant, Searches, Financial pages````, but I could see this being slowly applied all around not that it's there even in SearchKit. :fireworks: 

cc @christianwach @vingle

**Comment by shaneonabike** [comment]
@christianwach @vingle interesting I just noticed that Searchkit has this functionality as well (as of recent). I did a deep dive and found it declared [here](https://github.com/civicrm/civicrm-core/blob/7da5bd5b5ffde2bb1485c52825c832af5c29320f/ext/search_kit/css/crmSearchDisplayTable.css#L31). I think it would be good to generalize this here as I'm suggesting and then we could remove it from there. I noticed in Searchkit there is not switching for mobile/tablet sizes like I added here.

Finally, my goal is that after this is addition we can go through and simply replace all the ```sticky``` with ```crm-sticky```.  It's found in all these places

```
ext/legacycustomsearches/templates/CRM/Contact/Form/Search/Custom.tpl:            <thead class="sticky">
ext/legacycustomsearches/templates/CRM/Contact/Form/Search/Custom/Proximity.tpl:            <thead class="sticky">
ext/legacycustomsearches/templates/CRM/Contact/Form/Search/Custom/ContribSYBNT.tpl:            <thead class="sticky">
ext/legacycustomsearches/templates/CRM/Contact/Form/Search/Custom/ContributionAggregate.tpl:            <thead class="sticky">
ext/legacycustomsearches/templates/CRM/Contact/Form/Search/Custom/ActivitySearch.tpl:              <thead class="sticky">
ext/legacycustomsearches/templates/CRM/Contact/Form/Search/Custom/MultipleValues.tpl:            <thead class="sticky">
ext/legacycustomsearches/templates/CRM/Contact/Form/Search/Custom/EventDetails.tpl:            <thead class="sticky">
release-notes/5.29.0.md:- **Fix sticky table header on "Find Activities" page
release-notes/5.63.0.md:- **SearchKit - Add sticky table header style
release-notes/5.63.0.md:  Adds a "sticky" header option to SearchKit.
schema/Core/Note.entityType.php:    'icon' => 'fa-sticky-note',
templates/CRM/Financial/Page/BatchTransaction.tpl:  <thead class="sticky">
templates/CRM/Financial/Page/FinancialType.tpl:          <thead class="sticky">
templates/CRM/Financial/Page/FinancialTypeAccount.tpl:        <thead class="sticky">
templates/CRM/Financial/Page/FinancialAccount.tpl:        <thead class="sticky">
templates/CRM/Report/Form/Layout/Table.tpl:        {if !$sections} {* section headers and sticky headers aren't playing nice yet *}
templates/CRM/Report/Form/Layout/Table.tpl:            <thead class="sticky">
templates/CRM/Report/Form/Layout/Overlay.tpl:        {if !$sections} {* section headers and sticky headers aren't playing nice yet *}
templates/CRM/Report/Form/Contribute/DeferredRevenue.tpl:     <thead class="sticky">
templates/CRM/ACL/Form/WordPress/Permissions.tpl:<div class="crm-block crm-form-block crm-export-form-block crm-sticky">
templates/CRM/Core/Form/Task/Batch.tpl:         <thead class="sticky">
templates/CRM/Pledge/Form/Selector.tpl:    <thead class="sticky">
templates/CRM/Contact/Form/Selector.tpl:  <thead class="sticky">
templates/CRM/Contact/Form/Task/Batch.tpl:    <thead class="sticky">
templates/CRM/Contribute/Page/ContributionType.tpl:           <thead class="sticky">
templates/CRM/Contribute/Form/Selector.tpl:    <thead class="sticky">
templates/CRM/Contribute/Form/Task/Batch.tpl:    <thead class="sticky">
templates/CRM/Mailing/Page/Browse.tpl:      <thead class="sticky">
templates/CRM/Mailing/Page/Component.tpl:        <thead class="sticky">
templates/CRM/Mailing/Form/Selector.tpl:  <thead class="sticky">
templates/CRM/Campaign/Form/Selector.tpl:  <thead class="sticky">
templates/CRM/Price/Page/table.tpl:      <thead class="sticky">
templates/CRM/Price/Page/table.tpl:      <thead class="sticky">
templates/CRM/Price/Page/table.tpl:  <thead class="sticky">
templates/CRM/Activity/Form/Selector.tpl:   <thead class="sticky">
templates/CRM/Activity/Form/Task/Batch.tpl:    <thead class="sticky">
templates/CRM/Admin/Page/ParticipantStatusType.tpl:      <thead class="sticky">
templates/CRM/Admin/Page/MailSettings.tpl:        <thead class="sticky">
templates/CRM/Event/Form/Selector.tpl:<thead class="sticky">
templates/CRM/Event/Form/Task/Batch.tpl:       <thead class="sticky">
templates/CRM/Member/Page/MembershipStatus.tpl:        <thead class="sticky">
templates/CRM/Member/Form/Selector.tpl:<thead class="sticky">
templates/CRM/Member/Form/Task/Batch.tpl:         <thead class="sticky">

```

**Comment by vingle** [comment]
@shaneonabike - many more instances than I expected! 

How do you feel about combining this with looking at tables across Civi in general? 

The [DINO approach](https://lab.civicrm.org/dev/user-interface/-/wikis/CiviCRM-theming-in-the-age-of-SearchKit-&-FormBuilder-(DINO)) is: Document, Identify best version(s), New theme, Optimise markup - and that's how we updated all the Accordions.. ie found 8 patterns; decided a new better accordion pattern on a call; got the css support for it in core, then - the herculean work in Montreal optimising all the templates.

If all these table template files are going to be updated and tested, then maybe it'd good to make sure the entire table is the best Civi table it can be? Which would mean going back to [ThemeTest](https://lab.civicrm.org/extensions/themetest), seeing [if there are missing table patterns](https://lab.civicrm.org/extensions/themetest/-/tree/main/ang/themetest/snippets), agreeing (perhaps on a call) the best version (which might have variations, e.g. sortable, sticky header, alternating backgrounds, no borders, etc), and then updating the template files? I appreciate that's a bigger/longer job and process..

**Comment by christianwach** [comment]
> FWIW if you test this on mobile it might not work unless you modify the Viewport

@shaneonabike As I said on the CAU issue - this will need further though given that it might adversely affect WordPress admin if applied globally.

File: templates/CRM/ACL/Form/WordPress/Permissions.tpl [hunk 2]
```diff
@@ -12,42 +12,44 @@
...
+            <br/><span class="description">{$row.desc}</span>
+          {/if}
         </td>
-      {/foreach}
 
-      </tr>
-      {assign var="x" value=$x+1}
-    {/foreach}
+        {foreach from=$row.roles key=index item=role_name}
+          <td align="center" style="padding-top: 0.6em;">
```
Comment (vingle) [inline templates/CRM/ACL/Form/WordPress/Permissions.tpl:44]: Just noticed this inline styles here and at line 36 - can this be avoided?

Comment (shaneonabike) [reply templates/CRM/ACL/Form/WordPress/Permissions.tpl:44]: I think that was there before but I agree it's not ideal really.

File: css/civicrm.css [hunk 3]
```diff
@@ -125,17 +125,39 @@ input.crm-form-checkbox + label {
...
+.crm-sticky thead,
+.crm-container thead div.sticky-header {
   border-bottom: 2px solid #cfcec3;
   background-color: #fff;
   z-index: 10;
 }
 
+.crm-sticky thead {
+  position: sticky;
+  top: var(--crm-menubar-bottom);
```
Comment (vingle) [inline css/civicrm.css:146]: Should this be `--crm-sticky-top`? Otherwise I can't see this sit anywhere - tho then it would only be set in WordPress.

Comment (shaneonabike) [reply css/civicrm.css:146]: I haven't checked the other code actually. I was thinking about creating a master ticket (just haven't had time) about reworking all the tables. I have a few ideas of some things we can do to streamline things a bit. I'll grep (in about a week) for the class you mentioned and happy to change what I wrote to something already defined. I just didn't find anything before I think.

Comment (shaneonabike) [reply css/civicrm.css:146]: No I think I misunderstood your comment above and thought you were indicating that in fact it was a duplication. Wordpress CiviCRM generates this automatically based on the resizing of the window. The same variable exists for D9 (probably onward too). My thinking was to use the bottom, because with a sticky header it should essentially start at that bottom (below the actual menu) when the user gets to the point where it should be sticky.

Comment (vingle) [reply css/civicrm.css:146]: Ah I see! Had missed that Coleman [added that](https://github.com/civicrm/civicrm-core/blob/973f77a69233a23150cc5562ee1c19f68fe326cd/js/crm.menubar.js#L527) last year as a JS generated variable.

File: templates/CRM/ACL/Form/WordPress/Permissions.tpl (hunk 2 above)
Comment (colemanw) [reply templates/CRM/ACL/Form/WordPress/Permissions.tpl:44]: Probably shouldn't block the PR over this since it's pre-existing.

Comment (shaneonabike) [reply templates/CRM/ACL/Form/WordPress/Permissions.tpl:44]: I'm actually going to create a bigger WIP pertaining to all the tables and stuff but I just wanted to put this through so that I could tackle that after.

**Comment by shaneonabike** [comment]
Hey guys... any reason we can't merge this... it's been open for a while now

cc @vingle @colemanw

**Comment by vingle** [comment]
Looks good to me.
